// @wordpress/components/build/mobile/bottom-sheet/stepper-cell/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _cell = _interopRequireDefault(require("../cell"));

var _stepper = _interopRequireDefault(require("./stepper"));

var _style = _interopRequireDefault(require("./style.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var STEP_SPEED = 200;
var DEFAULT_STEP = 1;

var BottomSheetStepperCell =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(BottomSheetStepperCell, _Component);

  function BottomSheetStepperCell(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BottomSheetStepperCell);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(BottomSheetStepperCell).call(this, props));
    _this.announceValue = _this.announceValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onDecrementValue = _this.onDecrementValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onDecrementValuePressIn = _this.onDecrementValuePressIn.bind((0, _assertThisInitialized2.default)(_this));
    _this.onIncrementValue = _this.onIncrementValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onIncrementValuePressIn = _this.onIncrementValuePressIn.bind((0, _assertThisInitialized2.default)(_this));
    _this.onPressOut = _this.onPressOut.bind((0, _assertThisInitialized2.default)(_this));
    _this.startPressInterval = _this.startPressInterval.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(BottomSheetStepperCell, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
      clearTimeout(this.timeoutAnnounceValue);
    }
  }, {
    key: "onIncrementValue",
    value: function onIncrementValue() {
      var _this$props = this.props,
          step = _this$props.step,
          maxValue = _this$props.maxValue,
          onChangeValue = _this$props.onChangeValue,
          value = _this$props.value;
      var newValue = value + step;

      if (newValue <= maxValue) {
        onChangeValue(newValue);
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onDecrementValue",
    value: function onDecrementValue() {
      var _this$props2 = this.props,
          step = _this$props2.step,
          minValue = _this$props2.minValue,
          onChangeValue = _this$props2.onChangeValue,
          value = _this$props2.value;
      var newValue = value - step;

      if (newValue >= minValue) {
        onChangeValue(newValue);
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onIncrementValuePressIn",
    value: function onIncrementValuePressIn() {
      var _this2 = this;

      this.onIncrementValue();
      this.timeout = setTimeout(function () {
        _this2.startPressInterval(_this2.onIncrementValue);
      }, 500);
    }
  }, {
    key: "onDecrementValuePressIn",
    value: function onDecrementValuePressIn() {
      var _this3 = this;

      this.onDecrementValue();
      this.timeout = setTimeout(function () {
        _this3.startPressInterval(_this3.onDecrementValue);
      }, 500);
    }
  }, {
    key: "onPressOut",
    value: function onPressOut() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
    }
  }, {
    key: "startPressInterval",
    value: function startPressInterval(callback) {
      var _this4 = this;

      var counter = 0;
      this.interval = setInterval(function () {
        callback();
        counter += 1;

        if (counter === 10) {
          clearInterval(_this4.interval);

          _this4.startPressInterval(callback, STEP_SPEED / 2);
        }
      }, STEP_SPEED);
    }
  }, {
    key: "announceValue",
    value: function announceValue(value) {
      var label = this.props.label;

      if (_reactNative.Platform.OS === 'ios') {
        // On Android it triggers the accessibilityLabel with the value change
        clearTimeout(this.timeoutAnnounceValue);
        this.timeoutAnnounceValue = setTimeout(function () {
          _reactNative.AccessibilityInfo.announceForAccessibility("".concat(value, " ").concat(label));
        }, 300);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this5 = this;

      var _this$props3 = this.props,
          label = _this$props3.label,
          icon = _this$props3.icon,
          minValue = _this$props3.minValue,
          maxValue = _this$props3.maxValue,
          value = _this$props3.value,
          separatorType = _this$props3.separatorType;
      var isMinValue = value === minValue;
      var isMaxValue = value === maxValue;
      var accessibilityLabel = (0, _i18n.sprintf)(
      /* translators: accessibility text. Inform about current value. %1$s: Control label %2$s: Current value. */
      (0, _i18n.__)('%1$s. Current value is %2$s'), label, value);
      return (0, _element.createElement)(_reactNative.View, {
        accessible: true,
        accessibilityRole: "adjustable",
        accessibilityLabel: accessibilityLabel,
        accessibilityActions: [{
          name: 'increment'
        }, {
          name: 'decrement'
        }],
        onAccessibilityAction: function onAccessibilityAction(event) {
          switch (event.nativeEvent.actionName) {
            case 'increment':
              _this5.onIncrementValue();

              break;

            case 'decrement':
              _this5.onDecrementValue();

              break;
          }
        }
      }, (0, _element.createElement)(_cell.default, {
        accessibilityRole: "none",
        accessible: false,
        cellContainerStyle: _style.default.cellContainerStyles,
        cellRowContainerStyle: _style.default.cellRowStyles,
        disabled: true,
        editable: false,
        icon: icon,
        label: label,
        labelStyle: _style.default.cellLabel,
        separatorType: separatorType
      }, (0, _element.createElement)(_stepper.default, {
        isMaxValue: isMaxValue,
        isMinValue: isMinValue,
        onPressInDecrement: this.onDecrementValuePressIn,
        onPressInIncrement: this.onIncrementValuePressIn,
        onPressOut: this.onPressOut,
        value: value
      })));
    }
  }]);
  return BottomSheetStepperCell;
}(_element.Component);

BottomSheetStepperCell.defaultProps = {
  step: DEFAULT_STEP
};
var _default = BottomSheetStepperCell;
exports.default = _default;
