// reakit/lib/Form/FormGroup.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useSealedState');
require('../Id/IdState.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/useLiveRef');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isPlainObject');
var FormState = require('./FormState.js');
require('reakit-utils/isObject');
require('./utils/setAllIn.js');
require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
var getInputId = require('../getInputId-cbd3bbcc.js');
var getLabelId = require('../getLabelId-eba7e47b.js');
var shouldShowError = require('../shouldShowError-8cab5d01.js');
var Group = require('../Group/Group.js');

var unstable_useFormGroup = createHook.createHook({
  name: "FormGroup",
  compose: Group.useGroup,
  useState: FormState.unstable_useFormState,
  keys: ["name"],
  useProps: function useProps(options, htmlProps) {
    return _rollupPluginBabelHelpers._objectSpread2({
      id: getInputId.getInputId(options.name, options.baseId),
      tabIndex: -1,
      "aria-describedby": shouldShowError.getMessageId(options.name, options.baseId),
      "aria-labelledby": getLabelId.getLabelId(options.name, options.baseId),
      "aria-invalid": shouldShowError.shouldShowError(options, options.name)
    }, htmlProps);
  }
});
var unstable_FormGroup = createComponent.createComponent({
  as: "fieldset",
  useHook: unstable_useFormGroup
});

exports.unstable_FormGroup = unstable_FormGroup;
exports.unstable_useFormGroup = unstable_useFormGroup;
