// reakit/lib/Form/FormCheckbox.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
var React = require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useForkRef');
require('reakit-utils/isButton');
require('reakit-utils/tabbable');
require('reakit-utils/hasFocusWithin');
require('../Tabbable/Tabbable.js');
require('reakit-utils/removeIndexFromArray');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('../Checkbox/CheckboxState.js');
var Checkbox = require('../Checkbox/Checkbox.js');
require('../Id/IdState.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/useLiveRef');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isPlainObject');
var FormState = require('./FormState.js');
require('reakit-utils/isObject');
require('./utils/setAllIn.js');
var getIn = require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
var getInputId = require('../getInputId-cbd3bbcc.js');
var getLabelId = require('../getLabelId-eba7e47b.js');
var shouldShowError = require('../shouldShowError-8cab5d01.js');

var unstable_useFormCheckbox = createHook.createHook({
  name: "FormCheckbox",
  compose: Checkbox.useCheckbox,
  useState: FormState.unstable_useFormState,
  keys: ["name", "value"],
  useOptions: function useOptions(options, htmlProps) {
    var name = options.name || htmlProps.name;
    var value = typeof options.value !== "undefined" ? options.value : htmlProps.value;
    var state = getIn.unstable_getIn(options.values, name);

    var setState = function setState(val) {
      return options.update(name, val);
    };

    return _rollupPluginBabelHelpers._objectSpread2({}, options, {
      state: state,
      setState: setState,
      name: name,
      value: value
    });
  },
  useProps: function useProps(options, _ref) {
    var htmlOnBlur = _ref.onBlur,
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref, ["onBlur"]);

    var isBoolean = typeof options.value === "undefined";
    var onBlur = React.useCallback(function () {
      options.blur(options.name);
    }, [options.blur, options.name]);
    return _rollupPluginBabelHelpers._objectSpread2({
      "aria-invalid": shouldShowError.shouldShowError(options, options.name),
      name: getInputId.formatInputName(options.name),
      onBlur: useAllCallbacks.useAllCallbacks(onBlur, htmlOnBlur)
    }, isBoolean ? {
      id: getInputId.getInputId(options.name, options.baseId),
      "aria-describedby": shouldShowError.getMessageId(options.name, options.baseId),
      "aria-labelledby": getLabelId.getLabelId(options.name, options.baseId)
    } : {}, {}, htmlProps);
  }
});
var unstable_FormCheckbox = createComponent.createComponent({
  as: "input",
  useHook: unstable_useFormCheckbox
});

exports.unstable_FormCheckbox = unstable_FormCheckbox;
exports.unstable_useFormCheckbox = unstable_useFormCheckbox;
