// reakit/lib/Form/Form.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('../Id/IdState.js');
var IdGroup = require('../Id/IdGroup.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/useLiveRef');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isPlainObject');
var FormState = require('./FormState.js');
require('reakit-utils/isObject');
require('./utils/setAllIn.js');
require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');

var unstable_useForm = createHook.createHook({
  name: "Form",
  compose: IdGroup.unstable_useIdGroup,
  useState: FormState.unstable_useFormState,
  useProps: function useProps(options, _ref) {
    var htmlOnSubmit = _ref.onSubmit,
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref, ["onSubmit"]);

    var onSubmit = React.useCallback(function (event) {
      event.preventDefault();
      options.submit();
    }, [options.submit]);
    return _rollupPluginBabelHelpers._objectSpread2({
      role: "form",
      noValidate: true,
      onSubmit: useAllCallbacks.useAllCallbacks(onSubmit, htmlOnSubmit)
    }, htmlProps);
  }
});
var unstable_Form = createComponent.createComponent({
  as: "form",
  useHook: unstable_useForm
});

exports.unstable_Form = unstable_Form;
exports.unstable_useForm = unstable_useForm;
