// reakit/lib/Form/FormMessage.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
var Box = require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useSealedState');
require('../Id/IdState.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/useLiveRef');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isPlainObject');
var FormState = require('./FormState.js');
require('reakit-utils/isObject');
require('./utils/setAllIn.js');
var getIn = require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
require('../getInputId-cbd3bbcc.js');
var shouldShowError = require('../shouldShowError-8cab5d01.js');

function shouldShowMessage(_ref, name) {
  var touched = _ref.touched,
      messages = _ref.messages;
  return Boolean(getIn.unstable_getIn(touched, name) && getIn.unstable_getIn(messages, name));
}

var unstable_useFormMessage = createHook.createHook({
  name: "FormMessage",
  compose: Box.useBox,
  useState: FormState.unstable_useFormState,
  keys: ["name"],
  useProps: function useProps(options, htmlProps) {
    var children = shouldShowError.shouldShowError(options, options.name) ? getIn.unstable_getIn(options.errors, options.name) : undefined;
    children = children || (shouldShowMessage(options, options.name) ? getIn.unstable_getIn(options.messages, options.name) : undefined);
    return _rollupPluginBabelHelpers._objectSpread2({
      role: "alert",
      id: shouldShowError.getMessageId(options.name, options.baseId),
      children: children
    }, htmlProps);
  }
});
var unstable_FormMessage = createComponent.createComponent({
  as: "div",
  useHook: unstable_useFormMessage
});

exports.unstable_FormMessage = unstable_FormMessage;
exports.unstable_useFormMessage = unstable_useFormMessage;
