// reakit/lib/Form/FormInput.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useForkRef');
require('reakit-utils/isButton');
require('reakit-utils/tabbable');
require('reakit-utils/hasFocusWithin');
var Tabbable = require('../Tabbable/Tabbable.js');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('../Id/IdState.js');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/useLiveRef');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isPlainObject');
var FormState = require('./FormState.js');
require('reakit-utils/isObject');
require('./utils/setAllIn.js');
var getIn = require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
var getInputId = require('../getInputId-cbd3bbcc.js');
var getLabelId = require('../getLabelId-eba7e47b.js');
var shouldShowError = require('../shouldShowError-8cab5d01.js');

var unstable_useFormInput = createHook.createHook({
  name: "FormInput",
  compose: Tabbable.useTabbable,
  useState: FormState.unstable_useFormState,
  keys: ["name"],
  useOptions: function useOptions(options, _ref) {
    var name = _ref.name;
    return _rollupPluginBabelHelpers._objectSpread2({
      name: name
    }, options, {
      unstable_clickOnEnter: false,
      unstable_clickOnSpace: false
    });
  },
  useProps: function useProps(options, _ref2) {
    var htmlOnChange = _ref2.onChange,
        htmlOnBlur = _ref2.onBlur,
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref2, ["onChange", "onBlur"]);

    var onChange = React.useCallback(function (event) {
      options.update(options.name, event.target.value);
    }, [options.update, options.name]);
    var onBlur = React.useCallback(function () {
      options.blur(options.name);
    }, [options.blur, options.name]);
    return _rollupPluginBabelHelpers._objectSpread2({
      id: getInputId.getInputId(options.name, options.baseId),
      name: getInputId.formatInputName(options.name),
      value: getIn.unstable_getIn(options.values, options.name, ""),
      onChange: useAllCallbacks.useAllCallbacks(onChange, htmlOnChange),
      onBlur: useAllCallbacks.useAllCallbacks(onBlur, htmlOnBlur),
      "aria-describedby": shouldShowError.getMessageId(options.name, options.baseId),
      "aria-labelledby": getLabelId.getLabelId(options.name, options.baseId),
      "aria-invalid": shouldShowError.shouldShowError(options, options.name)
    }, htmlProps);
  }
});
var unstable_FormInput = createComponent.createComponent({
  as: "input",
  useHook: unstable_useFormInput
});

exports.unstable_FormInput = unstable_FormInput;
exports.unstable_useFormInput = unstable_useFormInput;
