// reakit/lib/Tab/TabPanel.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('reakit-utils/cx');
require('../Id/IdState.js');
require('../Id/IdGroup.js');
require('reakit-utils/useIsomorphicEffect');
require('../Disclosure/DisclosureState.js');
var DisclosureRegion = require('../Disclosure/DisclosureRegion.js');
require('../Rover/RoverState.js');
var __utils = require('../__utils-26bc192f.js');
var TabState = require('./TabState.js');

var useTabPanel = createHook.createHook({
  name: "TabPanel",
  compose: DisclosureRegion.useDisclosureRegion,
  useState: TabState.useTabState,
  keys: ["stopId"],
  useOptions: function useOptions(options) {
    return _rollupPluginBabelHelpers._objectSpread2({
      visible: options.selectedId === options.stopId
    }, options, {
      unstable_setBaseId: undefined
    });
  },
  useProps: function useProps(options, htmlProps) {
    return _rollupPluginBabelHelpers._objectSpread2({
      role: "tabpanel",
      tabIndex: 0,
      id: __utils.getTabPanelId(options.stopId, options.baseId),
      "aria-labelledby": __utils.getTabId(options.stopId, options.baseId)
    }, htmlProps);
  }
});
var TabPanel = createComponent.createComponent({
  as: "div",
  useHook: useTabPanel
});

exports.TabPanel = TabPanel;
exports.useTabPanel = useTabPanel;
