// reakit/es/Popover/index.js
import 'reakit-utils/warning';
import 'react';
import '../Id/IdProvider.js';
import 'reakit-system/createComponent';
import 'reakit-system/createHook';
import '../Box/Box.js';
import '../_rollupPluginBabelHelpers-f089acec.js';
import 'reakit-utils/useForkRef';
import 'reakit-utils/isButton';
import 'reakit-utils/tabbable';
import 'reakit-utils/hasFocusWithin';
import '../Tabbable/Tabbable.js';
import '../Button/Button.js';
import 'reakit-utils/useAllCallbacks';
import 'reakit-utils/useSealedState';
import 'reakit-system/useCreateElement';
import 'reakit-utils/usePipe';
import 'reakit-utils/cx';
import '../Id/IdState.js';
import '../Id/IdGroup.js';
import 'reakit-utils/useIsomorphicEffect';
import '../Disclosure/DisclosureState.js';
import '../Disclosure/DisclosureRegion.js';
import 'react-dom';
import '../Portal/Portal.js';
import '../Dialog/Dialog.js';
import 'body-scroll-lock';
import 'reakit-utils/useUpdateEffect';
import 'reakit-utils/closest';
import 'reakit-utils/removeItemFromArray';
import 'reakit-utils/getDocument';
import 'reakit-utils/useLiveRef';
import '../Dialog/DialogState.js';
import '../DialogBackdropContext-dd59a293.js';
import '../Dialog/DialogBackdrop.js';
import '../Disclosure/Disclosure.js';
import '../Dialog/DialogDisclosure.js';
import '@popperjs/core';
export { usePopoverState } from './PopoverState.js';
export { Popover, usePopover } from './Popover.js';
export { PopoverArrow, usePopoverArrow } from './PopoverArrow.js';
export { PopoverDisclosure, usePopoverDisclosure } from './PopoverDisclosure.js';
export { PopoverBackdrop, usePopoverBackdrop } from './PopoverBackdrop.js';
