// reakit/es/Popover/PopoverBackdrop.js
import 'reakit-utils/warning';
import 'react';
import '../Id/IdProvider.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import '../Box/Box.js';
import { _ as _objectWithoutPropertiesLoose, a as _objectSpread2 } from '../_rollupPluginBabelHelpers-f089acec.js';
import 'reakit-utils/useAllCallbacks';
import 'reakit-utils/useSealedState';
import 'reakit-utils/usePipe';
import 'reakit-utils/cx';
import '../Id/IdState.js';
import '../Id/IdGroup.js';
import 'reakit-utils/useIsomorphicEffect';
import '../Disclosure/DisclosureState.js';
import '../Disclosure/DisclosureRegion.js';
import 'react-dom';
import '../Portal/Portal.js';
import '../Dialog/DialogState.js';
import '../DialogBackdropContext-dd59a293.js';
import { useDialogBackdrop } from '../Dialog/DialogBackdrop.js';
import '@popperjs/core';
import { usePopoverState } from './PopoverState.js';

var usePopoverBackdrop = createHook({
  name: "PopoverBackdrop",
  compose: useDialogBackdrop,
  useState: usePopoverState,
  useOptions: function useOptions(_ref) {
    var _ref$modal = _ref.modal,
        modal = _ref$modal === void 0 ? false : _ref$modal,
        options = _objectWithoutPropertiesLoose(_ref, ["modal"]);

    return _objectSpread2({
      modal: modal
    }, options);
  }
});
var PopoverBackdrop = createComponent({
  as: "div",
  useHook: usePopoverBackdrop
});

export { PopoverBackdrop, usePopoverBackdrop };
