// reakit/lib/Id/Id.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var IdProvider = require('./IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
var Box = require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useSealedState');
var IdState = require('./IdState.js');

var unstable_useId = createHook.createHook({
  name: "Id",
  compose: Box.useBox,
  useState: IdState.unstable_useIdState,
  keys: ["id"],
  useOptions: function useOptions(options, htmlProps) {
    var generateId = React.useContext(IdProvider.unstable_IdContext);

    var _React$useState = React.useState(function () {
      // This comes from useIdState
      if (options.unstable_idCountRef) {
        options.unstable_idCountRef.current += 1;
        return "-" + options.unstable_idCountRef.current;
      } // If there's no useIdState, we check if `baseId` was passed (as a prop,
      // not from useIdState).


      if (options.baseId) {
        return "-" + generateId("");
      }

      return "";
    }),
        suffix = _React$useState[0]; // `baseId` will be the prop passed directly as a prop or via useIdState.
    // If there's neither, then it'll fallback to Context's generateId.
    // This generateId can result in a sequential ID (if there's a Provider)
    // or a random string (without Provider).


    var baseId = React.useMemo(function () {
      return options.baseId || generateId();
    }, [options.baseId, generateId]);
    var id = options.id || htmlProps.id || "" + baseId + suffix;
    return _rollupPluginBabelHelpers._objectSpread2({}, options, {
      id: id
    });
  },
  useProps: function useProps(options, htmlProps) {
    var id = typeof htmlProps.id === "undefined" ? options.id : htmlProps.id;
    return _rollupPluginBabelHelpers._objectSpread2({}, htmlProps, {
      id: id
    });
  }
});
var unstable_Id = createComponent.createComponent({
  as: "div",
  useHook: unstable_useId
});

exports.unstable_Id = unstable_Id;
exports.unstable_useId = unstable_useId;
