// reakit/lib/Disclosure/Disclosure.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('reakit-utils/warning');
require('react');
require('../Id/IdProvider.js');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
require('../Box/Box.js');
var _rollupPluginBabelHelpers = require('../_rollupPluginBabelHelpers-baac11a7.js');
require('reakit-utils/useForkRef');
require('reakit-utils/isButton');
require('reakit-utils/tabbable');
require('reakit-utils/hasFocusWithin');
require('../Tabbable/Tabbable.js');
var Button = require('../Button/Button.js');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('reakit-utils/useSealedState');
require('../Id/IdState.js');
require('reakit-utils/useIsomorphicEffect');
var DisclosureState = require('./DisclosureState.js');

var useDisclosure = createHook.createHook({
  name: "Disclosure",
  compose: Button.useButton,
  useState: DisclosureState.useDisclosureState,
  useProps: function useProps(options, _ref) {
    var htmlOnClick = _ref.onClick,
        ariaControls = _ref["aria-controls"],
        htmlProps = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_ref, ["onClick", "aria-controls"]);

    var controls = ariaControls ? ariaControls + " " + options.baseId : options.baseId;
    return _rollupPluginBabelHelpers._objectSpread2({
      onClick: useAllCallbacks.useAllCallbacks(options.toggle, htmlOnClick),
      "aria-expanded": Boolean(options.visible),
      "aria-controls": controls
    }, htmlProps);
  }
});
var Disclosure = createComponent.createComponent({
  as: "button",
  useHook: useDisclosure
});

exports.Disclosure = Disclosure;
exports.useDisclosure = useDisclosure;
