// reakit/es/Menu/MenuArrow.js
import 'reakit-utils/warning';
import 'react';
import '../Id/IdProvider.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import '../Box/Box.js';
import '../_rollupPluginBabelHelpers-f089acec.js';
import 'reakit-utils/useForkRef';
import 'reakit-utils/isButton';
import 'reakit-utils/tabbable';
import 'reakit-utils/hasFocusWithin';
import '../Tabbable/Tabbable.js';
import 'reakit-utils/useAllCallbacks';
import 'reakit-utils/useSealedState';
import '../Id/IdState.js';
import 'reakit-utils/useIsomorphicEffect';
import '../Disclosure/DisclosureState.js';
import 'reakit-utils/removeItemFromArray';
import '../Dialog/DialogState.js';
import 'reakit-utils/createOnKeyDown';
import '../Id/Id.js';
import '../Rover/RoverState.js';
import '../Rover/Rover.js';
import '@popperjs/core';
import '../Popover/PopoverState.js';
import { u as useMenuState } from '../MenuState-449bf74e.js';
import './MenuBarState.js';
import { usePopoverArrow } from '../Popover/PopoverArrow.js';

var useMenuArrow = createHook({
  name: "MenuArrow",
  compose: usePopoverArrow,
  useState: useMenuState
});
var MenuArrow = createComponent({
  as: "div",
  useHook: useMenuArrow
});

export { MenuArrow, useMenuArrow };
